shadowsocks-csharp/Controller/System/ProtocolHandler.cs

```diff
@@ -0,0 +1,104 @@
+﻿using Microsoft.Win32;
+using NLog;
+using Shadowsocks.Util;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Reflection;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace Shadowsocks.Controller
+{
+    static class ProtocolHandler
+    {
+        private static Logger logger = LogManager.GetCurrentClassLogger();
+
+        // Don't use Application.ExecutablePath
+        // see https://stackoverflow.com/questions/12945805/odd-c-sharp-path-issue
+        private static readonly string ExecutablePath = Assembly.GetEntryAssembly().Location;
+
+        // TODO: Elevate when necessary
+        public static bool Set(bool enabled)
+        {
+            RegistryKey ssURLAssociation = null;
+            try
+            {
+                ssURLAssociation = Registry.CurrentUser.CreateSubKey(@"SOFTWARE\Classes\ss", RegistryKeyPermissionCheck.ReadWriteSubTree);
+                if (ssURLAssociation == null)
+                {
+                    logger.Error(@"Failed to create HKCU\SOFTWARE\Classes\ss");
```

Comment:
More friendly message "Failed to create HKCU\SOFTWARE\Classes\Shadowssocks to register the URL association"

Reply:
> More friendly message "Failed to create HKCU\SOFTWARE\Classes\Shadowssocks to register the URL association"

I'll make a quick change to it. But just so you know, this operation shouldn't normally fail.